src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs

```diff
@@ -100,6 +101,43 @@ protected ImmutableArray<ISymbol> GetSymbols(IParameterSymbol parameter, int pos
                 .ToImmutableArray();
         }
 
+        private ImmutableArray<ITypeSymbol> SubstituteTypeParameters(ImmutableArray<ITypeSymbol> parameterTypeSymbols, SyntaxNode invocationExpression)
+        {
+            if (!parameterTypeSymbols.Any(t => t.IsKind(SymbolKind.TypeParameter)))
+            {
+                return parameterTypeSymbols;
+            }
+
+            var invocationSymbols = _context.SemanticModel.GetSymbolInfo(invocationExpression).GetAllSymbols();
+            if (invocationSymbols.Length == 0)
+            {
+                return parameterTypeSymbols;
+            }
+
+            var concreteTypes = ArrayBuilder<ITypeSymbol>.GetInstance();
+            foreach (var invocationSymbol in invocationSymbols)
+            {
+                var typeParameters = invocationSymbol.GetTypeParameters();
+                var typeArguments = invocationSymbol.GetTypeArguments();
+
+                foreach (var parameterTypeSymbol in parameterTypeSymbols)
+                {
+                    if (parameterTypeSymbol.IsKind<ITypeParameterSymbol>(SymbolKind.TypeParameter, out var typeParameter))
```

Comment:
what if the parameter type references a type parameter from the *outer type* not a method type parameter?  i think the lines below will crash.  please add tests.

Reply:
Added a comment for this.